tests/foreman/cli/test_leapp_client.py

```diff
@@ -0,0 +1,203 @@
+    cmd_out = target_sat.execute(
+        f"hammer repository list --search 'content_label ~ rhel-9' --content-view {c_view.name} "
+        f"--organization '{organization.name}' --lifecycle-environment '{lc_env.name}'"
+    )
+    assert cmd_out.status == 0
+    assert ("AppStream RPMs 9" in cmd_out.stdout) and ("BaseOS RPMs 9" in cmd_out.stdout)
+
+    # Preupgrade conditions and check
+    rhel_contenthost.run("rm -rf /root/tmp_leapp_py3")
+    # Remove directory if in-place upgrade already performed from RHEL7 to RHEL8
+    rhel_old_ver = rhel_contenthost.os_version
+    # 5. Update all packages and install Leapp utility
+    rhel_contenthost.run("yum clean all")
+    rhel_contenthost.run("yum repolist")
+
+    rhel_contenthost.run("subscription-manager release --set 8.8")
+    result = rhel_contenthost.run("dnf update -y")
+    assert result.status == 0
+    rhel_contenthost.run("yum install leapp-upgrade -y")
+
+    # Fixing inhibitors - download data files to avoid inhibitors
+    rhel_contenthost.run(
+        'curl -k "https://gitlab.cee.redhat.com/oamg/leapp-data/-/raw/stage/data/'
+        '{repomap.json,pes-events.json,device_driver_deprecation_data.json}"'
+        ' -o "/etc/leapp/files/#1"'
+    )
+    rhel_contenthost.run(
+        'sed -i "s/^AllowZoneDrifting=.*/AllowZoneDrifting=no/" /etc/firewalld/firewalld.conf'
+    )
+    # 6.1 Run LEAPP-PREUPGRADE Job Template-
+    template_id = (
+        target_sat.api.JobTemplate()
+        .search(query={'search': 'name="Run preupgrade via Leapp"'})[0]
+        .id
+    )
+    job = target_sat.api.JobInvocation().run(
+        synchronous=False,
+        data={
+            'job_template_id': template_id,
+            'targeting_type': 'static_query',
+            'search_query': f'name = {rhel_contenthost.hostname}',
+        },
+    )
+    target_sat.wait_for_tasks(
+        f'resource_type = JobInvocation and resource_id = {job["id"]}', poll_timeout=1800
+    )
+    result = target_sat.api.JobInvocation(id=job['id']).read()
+    assert result.succeeded == 1
+
+    # 6.2 Run LEAPP-UPGRADE Job Template-
+    template_id = (
+        target_sat.api.JobTemplate().search(query={'search': 'name="Run upgrade via Leapp"'})[0].id
+    )
+    job = target_sat.api.JobInvocation().run(
+        synchronous=False,
+        data={
+            'job_template_id': template_id,
+            'targeting_type': 'static_query',
+            'search_query': f'name = {rhel_contenthost.hostname}',
+            'inputs': {'Reboot': 'true'},
+        },
+    )
+    target_sat.wait_for_tasks(
+        f'resource_type = JobInvocation and resource_id = {job["id"]}', poll_timeout=1800
+    )
+    result = target_sat.api.JobInvocation(id=job['id']).read()
+    assert result.succeeded == 1
+    # Resume the rhel_contenthsot with ensure True to ping the virtual machine
+    rhel_contenthost.power_control(state=VmState.RUNNING, ensure=True)
```

Comment:
> Besides the typos in line 199, why do we need to `power_control` the VM? I see that reboot is set to true: `'inputs': {'Reboot': 'true'},`, therefore the VM should be running after the `Run upgrade via Leapp` is finished. Waking up the host if the assumption is that the host is up, does not seem to be right.

Host gets disconnected after reboot, in order to get back ssh connection using 'power_control' here.

> Could you create a new `ContentHost` method that would make use of the `connect` method and use it here? The method body will be similar to the contents of the following code snippet

I think it's not required to create a new `wait_for` method, because at line [#L196](https://github.com/SatelliteQE/robottelo/blob/845c03cbf9ac69ee492fb6458e00e5fb09c4e305/tests/foreman/cli/test_leapp_client.py#L196) script waiting till `leapp upgrade` & `reboot` completed, and right after that script loses ssh connection with host

Reply:
When I check at very first time, I observed once **LEAPP-UPGRADE Job Template** runs successfully and debugger moves to next line, rhel host was up & running and the script didn't have ssh connection with the host

Reply:
I get what your point is, but I believe that you just do not understand what I am asking you for.

You are correct, robottelo looses connection to the SSH daemon after the Leapp job reboots it. That is why you use `power_control`. You are taking an advantage of the fact that `power_control` has the `wait_for ContentHost.connect` logic already implemented. The reason why you should not use `power_control` is that it manipulates with the VM state. The VM should be running, therefore there is no need to ensure that it is really running, the Leapp job should reboot the host, not the test, nor the test should ensure the VM state.
Therefore the only thing that you need to do is to wait until `rhel_contenthost.connect()` succeeds.
And that is why I suggest to create a helper method that would encapsulate the wait logic.

Reply:
done